### PR TITLE
Add Throwable "cause" variants to warning log API

### PR DESCRIPTION
**Pull request by emanb29**
Adds variants to `log.warning` to bring it closer to `log.error`, in particular, adding a `cause: Throwable`.

A (not-so-hypothetical) use-case for this is an application for which there are a set of recoverable errors, which do not interfere with application functionality, but do indicate an error on the end-user's part. In such a case, it can be helpful to log a warning, and convenient to pass along the Throwable that caused said warning, in a way that is consistent with `log.error`

### CODE REVIEW

**Comment by octonato** [comment]
@emanb29, the build is failing because of bin-incompatibility with previous API. 

You will need to add a mima exclusion for 

```scala
ProblemFilters.exclude[ReversedMissingMethodProblem("akka.event.LoggingAdapter.notifyWarning")
```

You do that by adding a file similar to 
https://github.com/akka/akka/blob/master/akka-actor/src/main/mima-filters/2.6.14.backwards.excludes/30267-supervision-loglevel.excludes

Note that you will need add the file under `2.6.15.backwards.excludes` instead. 

Now, concerning the addition, I checked how logback and log4j do the same and indeed they do have overloaded variants accepting a `Throwable` for all log levels. 

I'm not sure what was the rationale behind the decision to not do the same in akka logging. Maybe something worth reconsider. Maybe others (@johanandren, @jrudolph, @patriknw) have more insight on why it wasn't added.

**Comment by emanb29** [comment]
Thanks @octonato! I made a copy/paste error but I found the travis command and am re-checking locally before updating the branch again.

**Comment by jrudolph** [comment]
> You will need to add a mima exclusion for
> 
> ```scala
> ProblemFilters.exclude[ReversedMissingMethodProblem("akka.event.LoggingAdapter.notifyWarning")
> ```

This might be an actual incompatibility, as this is a public class that might be implemented by third parties. If we want to include this, we might want to include a default implementation that calls the existing `notifyWarning` method (and throws away the Throwable).

> I'm not sure what was the rationale behind the decision to not do the same in akka logging. Maybe something worth reconsider. Maybe others (@johanandren, @jrudolph, @patriknw) have more insight on why it wasn't added.

I don't know but maybe it was just missed in the beginning and compatibility requirements prevented it afterwards?

Personally, I'm not against adding it when the compat issues can be resolved.

**Comment by octonato** [comment]
> This might be an actual incompatibility, as this is a public class that might be implemented by third parties. If we want to include this, we might want to include a default implementation that calls the existing notifyWarning method (and throws away the Throwable).

That's true, but I'm wondering which scenario you have in mind where a method addition could cause a problem. If I'm using such 3rd party library and I update my Akka version, the Akka version used by the 3rd party library will be evicted, no?

**Comment by johanandren** [comment]
The third party library could contain a subclass, that is not recompiled by upgrading Akka so does not have the new method. The third party library-case is in general the most problematic scenario with binary compatibility, compared to a users own leaf-project where sources are easier to recompile to fix a binary compatibility.

Not sure how likely that is in this case but we should step carefully. A default impl might work, but we should verify that.

**Comment by octonato** [comment]
> A default impl might work, but we should verify that.

I created a [small project](https://github.com/octonato/bin-compat-example) containing three sub-projects: `logger`, `logger-wrapper` and `app`. 

Adding a default impl for the new method as suggested by @jrudolph does the trick (as expected). 

Please, check readme file to understand how I tested it. I think this test covers the case here. Let me know if I missed something.

**Comment by johanandren** [comment]
Double checked the idea of default implementation with both 2.13 and 2.12 and that works fine (because Scalac turns the trait into an interface with `default` methods)

File: akka-actor/src/main/scala/akka/event/Logging.scala [hunk 1]
```diff
@@ -1199,6 +1199,7 @@ trait LoggingAdapter {
   protected def notifyError(message: String): Unit
   protected def notifyError(cause: Throwable, message: String): Unit
   protected def notifyWarning(message: String): Unit
+  protected def notifyWarning(cause: Throwable, message: String): Unit
```
Comment (johanandren) [inline akka-actor/src/main/scala/akka/event/Logging.scala:1202]: So the default implementation protecting custom LoggingAdapters would be:
```suggestion
  protected def notifyWarning(cause: Throwable, message: String): Unit = notifyWarning(message)
```

Not entirely sure MiMa will understand this is safe though. We might still need the filter.

File: akka-actor/src/main/scala/akka/event/Logging.scala [hunk 2]
```diff
@@ -1199,6 +1199,7 @@ trait LoggingAdapter {
   protected def notifyError(message: String): Unit
   protected def notifyError(cause: Throwable, message: String): Unit
   protected def notifyWarning(message: String): Unit
+  protected def notifyWarning(cause: Throwable, message: String): Unit = notifyWarning(message)
```
Comment (johanandren) [inline akka-actor/src/main/scala/akka/event/Logging.scala:1202]:
```suggestion
  protected def notifyWarning(@unused cause: Throwable, message: String): Unit = notifyWarning(message)
```

File: akka-actor/src/main/scala/akka/event/Logging.scala [hunk 3]
```diff
@@ -1981,6 +2025,8 @@ class BusLogging(val bus: LoggingBus, val logSource: String, val logClass: Class
     bus.publish(Error(cause, logSource, logClass, message, mdc))
   protected def notifyWarning(message: String): Unit =
     bus.publish(Warning(logSource, logClass, message, mdc))
+  protected def notifyWarning(cause: Throwable, message: String): Unit =
```
Comment (johanandren) [inline akka-actor/src/main/scala/akka/event/Logging.scala:2028]:
```suggestion
  override protected def notifyWarning(cause: Throwable, message: String): Unit =
```

File: akka-actor/src/main/mima-filters/2.6.15.backwards.excludes/30389-logging-adapter-notify.excludes [hunk 4]
```diff
@@ -0,0 +1,2 @@
+# additional overload added
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.event.LoggingAdapter.notifyWarning")
```
Comment (johanandren) [inline akka-actor/src/main/mima-filters/2.6.15.backwards.excludes/30389-logging-adapter-notify.excludes:2]: I tried it out and it looks like we can delete this exclude now and MiMa is still happy

File: akka-actor/src/main/scala/akka/event/Logging.scala [hunk 5]
```diff
@@ -1981,6 +2025,8 @@ class BusLogging(val bus: LoggingBus, val logSource: String, val logClass: Class
...
+    bus.publish(Warning(cause, logSource, logClass, message, mdc))
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/event/Logging.scala:2029]: Funny that we had prepared for it with that `Warning4` class, but not exposing it in the LoggingAdapter. Must have been an oversight.

**Review by patriknw — COMMENTED** [review]
looking good

**Comment by patriknw** [comment]
Completed in https://github.com/akka/akka/pull/31525

**Comment by patriknw** [comment]
@emanb29 Could you complete this in this original PR instead? Github is stupid in the way that it will change the author to me if we merge it from my PR. What I did was to remove the mima exclude files and added some `@nowarn` and `@override` that the compiler complained about. See https://github.com/akka/akka/pull/31525

**Comment by He-Pin** [comment]
Maybe submit a pr to his branch ？

**Comment by patriknw** [comment]
Merged to main as 044bb2dc0a3804d7fe33a1d3e988830256c108fa

**Pull request closed without merging**